### PR TITLE
cordova7_post : cordova7 release post

### PR DESCRIPTION
**Pull request by audreyso**
Not ready to merge yet, but would like some advice/feedback early on!

### What does this PR do?
Blog post for cordova@7 release
- Not ready to merge yet, but would appreciate some advice/feedback early on if anything is missing or incorrect! Thank you! :)

### Checklist
- [X] [Reported an issue](http://cordova.apache.org/contribute/issues.html) in the JIRA database
- [X] Commit message follows the format: "CB-3232: (android) Fix bug with resolving file paths", where CB-xxxx is the JIRA ID & "android" is the platform affected.
- [X] Added automated test coverage as appropriate for this change.

### CODE REVIEW

File: www/_posts/2017-04-21-cordova-7.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+* Users can install any platform due to removing known platform check for platform API. For instance, users can now add custom platforms to their projects.
+* Platforms and plugins are now required to have a package.json file.
+
+Example: 
+
+	cordova platform add custom-platform-name
+
+Link to how to create your own platforms:
+
+* A cordova config command has been created to `set`, `get`, `delete`, `edit`, and `list` global cordova options. For example, you can use the following command `cordova config set <key> <value>` to set the value of autosave or fetch to true or false.
```
Comment (shazron) [inline www/_posts/2017-04-21-cordova-7.md]: `cordova config` in backticks at the start of the sentence

File: www/_posts/2017-04-21-cordova-7.md [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+In this case, you are turning off autosave as the default setting. Example: 
+
+	cordova config set autosave false
+
+In this case, you are turning off fetch as the default setting. Example:
+
+	cordova config set fetch false
+
+In addition, the cordova config command supports browserify setting, which allows the JavaScript of plugins to be loaded at build time compared to run time. For instance, if the browserify value is not explicility passed in by the user, the cordova config command will automatically set the browserify value saved in `~/.config/configstore/` to saved `globally`. Users can get and set browserify. Example:
```
Comment (shazron) [inline www/_posts/2017-04-21-cordova-7.md]: `cordova config` in backticks ( 2 references). 
"the cordova config command supports browserify setting" should have a definite article before "browserify".
"explicility" (spelling)
"to saved `globally`" <-- grammar? unclear

File: www/_posts/2017-04-21-cordova-7.md [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+* Fetch is now the default setting for fetching platforms. Fetch uses your system `npm` to `npm install` modules into your project. The `--fetch` flag is no longer required. Use the `--nofetch` flag to use the older cordova fetching method (pre `cordova@7`) and prevent using `npm install` to do the fetching.
+* After `cordova prepare` is run, `package.json` and `config.xml` should contain identical platforms and versions.  In case of conflicts, `package.json` is given precedence over `config.xml`. For example, suppose `package.json` contains `cordova-android@6.0.0` and `config.xml` contains `cordova-android@4.0.0`. After `cordova prepare` is run, `config.xml` and `package.json` will each contain only `cordova-android@6.0.0`.
+* Users can install any platform due to removing known platform check for platform API. For instance, users can now add custom platforms to their projects.
```
Comment (shazron) [inline www/_posts/2017-04-21-cordova-7.md]: "due to removing known platform check for platform API" <-- grammar

File: www/_posts/2017-04-21-cordova-7.md [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+categories: news
+tags: news releases
+---
+
+We are happy to announce that `Apache Cordova 7.0.0` has been released!
+
+Most notable changes include:
+* If a `package.json` does not exist in your project, it will be auto-created for you when `cordova prepare` is called.
+* When adding a platform or plugin, it will automatically save that platform or plugin to your `config.xml` and `package.json`. Details about platform and plugin versions are also automatically saved in `config.xml` and `package.json`. The `--save` flag is no longer required to save. Use `--nosave` to prevent saving to `config.xml` or `package.json`.
+* Fetch is now the default setting for fetching platforms. Fetch uses your system `npm` to `npm install` modules into your project. The `--fetch` flag is no longer required. Use the `--nofetch` flag to use the older cordova fetching method (pre `cordova@7`) and prevent using `npm install` to do the fetching.
```
Comment (shazron) [inline www/_posts/2017-04-21-cordova-7.md]: "prevent using `npm install` to do the fetching." -> "prevent the use of `npm install` to do the fetching."

File: www/_posts/2017-04-21-cordova-7.md (hunk 3 above)
Comment (purplecabbage) [inline www/_posts/2017-04-21-cordova-7.md]: ~* Users can install any platform due to removing known platform check for platform API. For instance, users can now add custom platforms to their projects.~
* We have removed a requirement that `platform add` be called with a '_known platform_' and now simply expect that whatever you pass '_behaves like a known platform_'.  This will allow future custom platforms, and modified version of existing platforms.

**Comment by purplecabbage** [comment]
👍

File: www/_posts/2017-04-21-cordova-7.md [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+title:  "Apache Cordova 7.0.0"
+categories: news
+tags: news releases
+---
+
+We are happy to announce that `Apache Cordova 7.0.0` has been released!
+
+Most notable changes include:
+* If a `package.json` does not exist in your project, it will be auto-created for you when `cordova prepare` is called.
+* When adding a platform or plugin, it will automatically save that platform or plugin to your `config.xml` and `package.json`. Details about platform and plugin versions are also automatically saved in `config.xml` and `package.json`. The `--save` flag is no longer required to save. Use `--nosave` to prevent saving to `config.xml` or `package.json`.
```
Comment (filmaj) [inline www/_posts/2017-04-21-cordova-7.md]: Perhaps I would reword this to say that the previous default behaviour of not-saving has changed to saving? For me, when reading release notes, I want to understand how the new release impacts my use of the program in comparison to the prior release, so I appreciate calling out differences in default behaviour.

File: www/_posts/2017-04-21-cordova-7.md [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+* Fetch is now the default setting for fetching platforms. Fetch uses your system `npm` to `npm install` modules into your project. The `--fetch` flag is no longer required. Use the `--nofetch` flag to use the older cordova fetching method (pre `cordova@7`) and prevent using `npm install` to do the fetching.
+* After `cordova prepare` is run, `package.json` and `config.xml` should contain identical platforms and versions.  In case of conflicts, `package.json` is given precedence over `config.xml`. For example, suppose `package.json` contains `cordova-android@6.0.0` and `config.xml` contains `cordova-android@4.0.0`. After `cordova prepare` is run, `config.xml` and `package.json` will each contain only `cordova-android@6.0.0`.
+* Users can install any platform due to removing known platform check for platform API. For instance, users can now add custom platforms to their projects.
+* Platforms and plugins are now required to have a package.json file.
```
Comment (filmaj) [inline www/_posts/2017-04-21-cordova-7.md]: Backticks on `package.json` here, like previous references?

File: www/_posts/2017-04-21-cordova-7.md [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+In addition, the cordova config command supports browserify setting, which allows the JavaScript of plugins to be loaded at build time compared to run time. For instance, if the browserify value is not explicility passed in by the user, the cordova config command will automatically set the browserify value saved in `~/.config/configstore/` to saved `globally`. Users can get and set browserify. Example:
+	
+	cordova config get browserify
+	cordova config set browserify true
+
+* Removed support for node 0.x. Only node versions 4.0 and up are supported.
+* Removed support for `cordova-windows@8` and under.
+* Removed support for `cordova-android@4` and under.
+* Removed support for `cordova-ios@3` and under.
```
Comment (filmaj) [inline www/_posts/2017-04-21-cordova-7.md]: Hmm, is "and under" the right word to use when talking about version numbers? Maybe "and lower"?

Comment (purplecabbage) [inline www/_posts/2017-04-21-cordova-7.md]: and older?

File: www/_posts/2017-04-21-cordova-7.md (hunk 4 above)
Comment (kerrishotts) [inline www/_posts/2017-04-21-cordova-7.md]: Perhaps: use the `--nofetch` flag to revert to pre-7.0 behavior (`npm install` is _not_ used to fetch modules)

Comment (kerrishotts) [inline www/_posts/2017-04-21-cordova-7.md]: Also: if leaving Cordova in, it needs capitalized.

File: www/_posts/2017-04-21-cordova-7.md (hunk 1 above)
Comment (kerrishotts) [inline www/_posts/2017-04-21-cordova-7.md]: to set the value of `autosave` or `fetch` to `true` or `false`.

Comment (kerrishotts) [inline www/_posts/2017-04-21-cordova-7.md]: global cordova --> global Cordova

File: www/_posts/2017-04-21-cordova-7.md (hunk 2 above)
Comment (kerrishotts) [inline www/_posts/2017-04-21-cordova-7.md]: ?: In this case --> In the following case

or, "For example, if you want to turn off `fetch` as the default setting:"

To me, "this" refers to something that has already been mentioned, so I was looking above, not below.

Also, backticks around `fetch`, `autosave`, `browserify`

File: www/_posts/2017-04-21-cordova-7.md (hunk 4 above)
Comment (audreyso) [inline www/_posts/2017-04-21-cordova-7.md]: Thanks @kerrishotts Would I use backticks around fetch here --> "Fetch is now the default setting for fetching platforms. Fetch uses your system `npm` to `npm install` modules into your project."

Or only below where you mentioned it?

File: www/_posts/2017-04-21-cordova-7.md (hunk 7 above)
Comment (stevengill) [inline www/_posts/2017-04-21-cordova-7.md]: we also dropped support for older osx versions now

Comment (audreyso) [inline www/_posts/2017-04-21-cordova-7.md]: What version should be here?
Removed support for `cordova-osx@??` and older.

Comment (stevengill) [inline www/_posts/2017-04-21-cordova-7.md]: check out the deprecated field in platformConfigs.json. It should be equal to whenever osx got api.js support

File: www/_posts/2017-04-21-cordova-7.md (hunk 3 above)
Comment (imhotep) [inline www/_posts/2017-04-21-cordova-7.md]: @audreyso @purplecabbage I still don't understand what you guys mean exactly. People who don't work on Cordova don't know about platform API or "known" platforms. Could it just be this?
  - Support for custom platforms.

Comment (purplecabbage) [inline www/_posts/2017-04-21-cordova-7.md]: Yeah, that's probably better ... it is somewhat of an incomplete feature right now as it requires substantial documentation ( how to write a platform ) and some thought around plugin support.

Comment (audreyso) [inline www/_posts/2017-04-21-cordova-7.md]: @purplecabbage @imhotep How is this --> "We have added support for custom platforms. This will allow future custom platforms, and modified versions of existing platforms." or just "Support for custom platforms."

File: www/_posts/2017-04-21-cordova-7.md (hunk 4 above)
Comment (kerrishotts) [inline www/_posts/2017-04-21-cordova-7.md]: Hmmm. To me you're referring to the feature named "fetch", not the setting (although you do use the word "setting" -- I think I would use "method"). So:

"Fetch is now the default method for fetching platforms. Fetch uses your system...."

You could reword this a bit to avoid the issue entirely, but I think this is clear enough without being overly verbose.

File: www/_posts/2017-04-21-cordova-7.md (hunk 3 above)
Comment (imhotep) [inline www/_posts/2017-04-21-cordova-7.md]: @audreyso I understand both of those sentences so I would say pick whichever one you like.

File: www/_posts/2017-04-21-cordova-7.md [hunk 8]
```diff
@@ -12,35 +12,37 @@ We are happy to announce that `Apache Cordova 7.0.0` has been released!
...
-* Removed support for node 0.x. Only node versions 4.0 and up are supported.
-* Removed support for `cordova-windows@8` and under.
-* Removed support for `cordova-android@4` and under.
-* Removed support for `cordova-ios@3` and under.
+* Dropped support for node 0.x. Only node versions 4.0 and up are supported.
+* Dropped support for wp8 (windows phone 8).
+* `cordova-windows` : dropped support for any versions older than 4.0.0.
+* `cordova-android` : dropped support for any versions older than 5.0.0.
+* `cordova-ios` : dropped support for any versions older than 4.0.0.
+* `cordova-osx` : dropped support for any versions older than 4.0.0.
```
Comment (audreyso) [inline www/_posts/2017-04-21-cordova-7.md]: @stevengill are these correct? just want to make sure!

Comment (stevengill) [inline www/_posts/2017-04-21-cordova-7.md]: yup!

File: www/_posts/2017-04-21-cordova-7.md [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+* Fetch is now the default method for fetching platforms. Fetch uses your system `npm` to `npm install` modules into your project. The `--fetch` flag is no longer required. Use the `--nofetch` flag to revert to pre-`Cordova@7.0` behavior (`npm install` is not used to fetch modules).
+* After `cordova prepare` is run, `package.json` and `config.xml` should contain identical platforms and versions.  In case of conflicts, `package.json` is given precedence over `config.xml`. For example, suppose `package.json` contains `cordova-android@6.0.0` and `config.xml` contains `cordova-android@4.0.0`. After `cordova prepare` is run, `config.xml` and `package.json` will each contain only `cordova-android@6.0.0`.
+* We have added support for custom platforms. This will allow future custom platforms, and modified versions of existing platforms.
+* Platforms and plugins are now required to have a `package.json` file.
```
Comment (dpogue) [inline www/_posts/2017-04-21-cordova-7.md]: This line looks like it got inserted in-between the line above and the example below. I'd suggest moving it above the preceding bullet point.

File: www/_posts/2017-04-21-cordova-7.md [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+
+Example:
```
Comment (dpogue) [inline www/_posts/2017-04-21-cordova-7.md]: Indent by 2 to keep it nested under the bullet item

File: www/_posts/2017-04-21-cordova-7.md [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+* Fetch is now the default method for fetching platforms. Fetch uses your system `npm` to `npm install` modules into your project. The `--fetch` flag is no longer required. Use the `--nofetch` flag to revert to pre-`Cordova@7.0` behavior (`npm install` is not used to fetch modules).
+* After `cordova prepare` is run, `package.json` and `config.xml` should contain identical platforms and versions.  In case of conflicts, `package.json` is given precedence over `config.xml`. For example, suppose `package.json` contains `cordova-android@6.0.0` and `config.xml` contains `cordova-android@4.0.0`. After `cordova prepare` is run, `config.xml` and `package.json` will each contain only `cordova-android@6.0.0`.
+* We have added support for custom platforms. This will allow future custom platforms, and modified versions of existing platforms.
+* Platforms and plugins are now required to have a `package.json` file.
+
+Example: 
+
+	cordova platform add custom-platform-name
+
+Link to how to create your own platforms:
```
Comment (dpogue) [inline www/_posts/2017-04-21-cordova-7.md]: Indent by 2 to keep it nested under the bullet item

File: www/_posts/2017-04-21-cordova-7.md [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+
+* A `cordova config` command has been created to `set`, `get`, `delete`, `edit`, and `list` global Cordova options. For example, you can use the following command `cordova config set <key> <value>` to set the value of `autosave` or `fetch` to `true` or `false`. 
+
+ In the following case, you are turning off `autosave` as the default setting. Example:
```
Comment (dpogue) [inline www/_posts/2017-04-21-cordova-7.md]: Indent with 2 spaces to keep it nested under the bullet item

File: www/_posts/2017-04-21-cordova-7.md [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+
+Link to how to create your own platforms:
+
+* A `cordova config` command has been created to `set`, `get`, `delete`, `edit`, and `list` global Cordova options. For example, you can use the following command `cordova config set <key> <value>` to set the value of `autosave` or `fetch` to `true` or `false`. 
+
+ In the following case, you are turning off `autosave` as the default setting. Example: 
+
+	cordova config set autosave false
+
+In the following case, you are turning off `fetch` as the default setting. Example:
```
Comment (dpogue) [inline www/_posts/2017-04-21-cordova-7.md]: Indent by 2

File: www/_posts/2017-04-21-cordova-7.md [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+
+	cordova config set fetch false
+
+In addition, the `cordova config` command supports the `browserify` setting, which allows the JavaScript of plugins to be loaded at build time compared to run time. For instance, if the `browserify` value is not explicility passed in by the user, the `cordova config` command will automatically set the `browserify` value saved in `~/.config/configstore/` to be saved `globally`. Users can get and set `browserify`. Example:
```
Comment (dpogue) [inline www/_posts/2017-04-21-cordova-7.md]: indent by 2

File: www/_posts/2017-04-21-cordova-7.md [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+
+Most notable changes include:
+
+* [cordova@7.0.0] (https://www.npmjs.org/package/cordova)
+* [cordova-lib@7.0.0] (https://www.npmjs.org/package/cordova-lib)
+* [plugman@1.5.0] (https://www.npmjs.com/package/plugman)
+* [cordova-app-hello-world@3.12.0] (https://www.npmjs.com/package/cordova-app-hello-world)
+* [cordova-common@2.0.3] (https://www.npmjs.com/package/cordova-common)
+* [cordova-create@1.1.0] (https://www.npmjs.com/package/cordova-create)
+* [cordova-fetch@1.1.0] (https://www.npmjs.com/package/cordova-fetch)
```
Comment (dpogue) [inline www/_posts/2017-04-21-cordova-7.md]: No space between the `[text]` and the `(href)`, otherwise it doesn't link properly :)

File: www/_posts/2017-04-21-cordova-7.md [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+* Fetch is now the default method for fetching platforms. Fetch uses your system `npm` to `npm install` modules into your project. The `--fetch` flag is no longer required. Use the `--nofetch` flag to revert to pre-`Cordova@7.0` behavior (`npm install` is not used to fetch modules).
+* After `cordova prepare` is run, `package.json` and `config.xml` should contain identical platforms and versions.  In case of conflicts, `package.json` is given precedence over `config.xml`. For example, suppose `package.json` contains `cordova-android@6.0.0` and `config.xml` contains `cordova-android@4.0.0`. After `cordova prepare` is run, `config.xml` and `package.json` will each contain only `cordova-android@6.0.0`.
+* Platforms and plugins are now required to have a `package.json` file.
+* We have added support for custom platforms. This will allow future custom platforms, and modified versions of existing platforms.
+
+	Example: 
+
+	cordova platform add custom-platform-name
```
Comment (dpogue) [inline www/_posts/2017-04-21-cordova-7.md]: This (and the ones in the following paragraphs) aren't getting rendered as code blocks. You probably need to surround them with 3 backticks to force it.

File: www/_posts/2017-04-21-cordova-7.md [hunk 17]
```diff
@@ -0,0 +1,146 @@
...
+
+* [cordova@7.0.0](https://www.npmjs.org/package/cordova)
+* [cordova-lib@7.0.0](https://www.npmjs.org/package/cordova-lib)
+* [plugman@1.5.0](https://www.npmjs.com/package/plugman)
+* [cordova-app-hello-world@3.12.0](https://www.npmjs.com/package/cordova-app-hello-world)
+* [cordova-common@2.0.3](https://www.npmjs.com/package/cordova-common)
+* [cordova-create@1.1.0](https://www.npmjs.com/package/cordova-create)
+* [cordova-fetch@1.1.0](https://www.npmjs.com/package/cordova-fetch)
+
+* If a `package.json` does not exist in your project, it will be auto-created for you when `cordova prepare` is called.
```
Comment (piotrowski) [inline www/_posts/2017-04-21-cordova-7.md]: Is this part or first step of a bigger change? Like moving away from `config.xml` for platform and plugin management to using `package.json` for all that?

File: www/_posts/2017-04-21-cordova-7.md [hunk 18]
```diff
@@ -0,0 +1,146 @@
...
+* When adding a platform or plugin, the previous default behavior of not-saving has changed to saving. Now it will automatically save that platform or plugin to your `config.xml` and `package.json`. Details about platform and plugin versions are also automatically saved in `config.xml` and `package.json`. The `--save` flag is no longer required to save. Use `--nosave` to prevent saving to `config.xml` or `package.json`.
+* Fetch is now the default method for fetching platforms. Fetch uses your system `npm` to `npm install` modules into your project. The `--fetch` flag is no longer required. Use the `--nofetch` flag to revert to pre-`Cordova@7.0` behavior (`npm install` is not used to fetch modules).
```
Comment (piotrowski) [inline www/_posts/2017-04-21-cordova-7.md]: Similar question here: Is this part of a bigger change? How did getting platforms work before? Why does this change?

File: www/_posts/2017-04-21-cordova-7.md (hunk 17 above)
Comment (stevengill) [inline www/_posts/2017-04-21-cordova-7.md]: yes. But for cordova-7 we will be support both config.xml & package.json. The goal is to move over to `package.json` for platform/plugin management completely.

Comment (stevengill) [inline www/_posts/2017-04-21-cordova-7.md]: Giving downstream distributions (like visual studio) time to migrate over to package.json

File: www/_posts/2017-04-21-cordova-7.md (hunk 18 above)
Comment (stevengill) [inline www/_posts/2017-04-21-cordova-7.md]: previously, we did a `npm cache add` to a global directory and then copied over the contents. Now we shell out to npm, so we get npm's ability to save to package.json & npm's dependency fetching/resolution out of the box. 

`npm cache add` didn't `npm install`. So we had to bundle node_modules directory for all of our platforms. Now we don't have to bundle that directory anymore either.

Comment (piotrowski) [inline www/_posts/2017-04-21-cordova-7.md]: Excuse my ignorance, but what does "bunde that directory" mean? 
Does this in effect mean the platforms are now normal node_modules that can be handled like normal libraries?

**Comment by piotrowski** [comment]
Thanks @stevengill for the responses on my comments, all makes sense. I also noticed I am a bit late as the blog post is already published.

But: This is missing from the blog post! Right now the post is a very technical listing of stuff, just telling us what has changed but missing the why. Adding this as an explanation why this is happening would be great.

**Comment by filmaj** [comment]
Perhaps best left for a follow-up blog post? I think the "why" behind the changes is important, as @piotrowski points out.
